net/art/table.go

```diff
@@ -16,147 +16,624 @@ import (
+			// This strideTable contains a route that may be relevant to our
+			// search, remember it.
+			stridePrefixes[strideIdx] = st.prefix
+			strideRoutes[strideIdx] = rt
+			strideIdx++
 		}
 		if child == nil {
-			// No sub-routes further down, whatever we have recorded in ret is
-			// the result.
-			return ret
+			// No sub-routes further down, the last thing we recorded
+			// in strideRoutes is tentatively the result, barring
+			// misdirection from path compression.
+			break findLeaf
 		}
 		st = child
+		// Path compression means we may be skipping over some intermediate
+		// tables. We have to skip forward to whatever depth st now references.
+		i = st.prefix.Bits() / 8
 	}
 
-	// Unreachable because Insert/Delete won't allow the leaf strideTables to
-	// have children, so we must return via the nil check in the loop.
-	panic("unreachable")
+	// Walk backwards through the hits we recorded in strideRoutes and
+	// stridePrefixes, returning the first one whose subtree matches addr.
+	//
+	// In the common case where path compression did not mislead us, we'll
+	// return on the first loop iteration because the last route we recorded was
+	// the correct most-specific route.
+	for strideIdx > 0 {
+		strideIdx--
+		if stridePrefixes[strideIdx].Contains(addr) {
+			return strideRoutes[strideIdx]
+		}
+	}
+
+	// We either found no route hits at all (both previous loops terminated
+	// immediately), or we went on a wild goose chase down a compressed path for
+	// the wrong prefix, and also found no usable routes on the way back up to
+	// the root. This is a miss.
+	return nil
 }
 
 // Insert adds pfx to the table, with value val.
 // If pfx is already present in the table, its value is set to val.
 func (t *Table[T]) Insert(pfx netip.Prefix, val *T) {
+	t.init()
 	if val == nil {
 		panic("Table.Insert called with nil value")
 	}
+
+	// The standard library doesn't enforce normalized prefixes (where
+	// the non-prefix bits are all zero). These algorithms require
+	// normalized prefixes, so do it upfront.
+	pfx = pfx.Masked()
+
+	if debugInsert {
+		defer func() {
+			fmt.Printf("%s", t.debugSummary())
+		}()
+		fmt.Printf("\ninsert: start pfx=%s\n", pfx)
+	}
+
 	st := &t.v4
 	if pfx.Addr().Is6() {
 		st = &t.v6
 	}
+
+	// This algorithm is full of off-by-one headaches that boil down
+	// to the fact that pfx.Bits() has (2^n)+1 values, rather than
+	// just 2^n. For example, an IPv4 prefix length can be 0 through
+	// 32, which is 33 values.
+	//
+	// This extra possible value creates a lot of problems as we do
+	// bits and bytes math to traverse strideTables below. So, we
+	// treat the default route 0/0 specially here, that way the rest
+	// of the logic goes back to having 2^n values to reason about,
+	// which can be done in a nice and regular fashion with no edge
+	// cases.
+	if pfx.Bits() == 0 {
+		if debugInsert {
+			fmt.Printf("insert: default route\n")
+		}
+		st.insert(0, 0, val)
+		return
+	}
+
 	bs := pfx.Addr().AsSlice()
```

Comment:
can you move this down near the for loop?

Reply:
Done.